core/dbt/context/providers.py

```diff
@@ -1382,7 +1382,8 @@ def sql(self) -> Optional[str]:
         if self.model.language == ModelLanguage.sql:  # type: ignore[union-attr]
             # If the model is deferred and the adapter doesn't support zero-copy cloning, then select * from the prod
             # relation
-            if getattr(self.model, "defer_relation", None):
+            # TODO: avoid routing on args.which if possible
+            if getattr(self.model, "defer_relation", None) and self.config.args.which == "clone":
```

Comment:
Definitely want to avoid the need for this. I had previously wondered if we'd be able to consolidate the deferral methods used by `clone` and other commands, by:
- clobbering unselected resources with the production versions (to support "traditional" deferral)
- adding `defer_relation` for selected resources only (to support `clone`)

(this comment/issue: https://github.com/dbt-labs/dbt-core/issues/7965)

Reply:
Discussing live with @MichelleArk: We should overwrite `compiled_code` (also aliased as `sql`) within the `clone` task. Then we don't need to have this task-aware conditional logic live in the context method.